GVFS/GVFS.Common/FileBasedCollections/BinaryFileBasedCollection.cs

```diff
@@ -0,0 +1,450 @@
+        /// <remarks>If retryUntilSuccess is true, OpenOrCreateDataFile will only attempt to retry when the error is non-fatal</remarks>
+        private void OpenOrCreateDataFile(bool retryUntilSuccess)
+        {
+            int attempts = 0;
+            Exception lastException = null;
+            while (true)
+            {
+                try
+                {
+                    if (this.dataFileHandle == null)
+                    {
+                        this.dataFileHandle = this.fileSystem.OpenFileStream(
+                            this.DataFilePath,
+                            FileMode.OpenOrCreate,
+                            this.collectionAppendsDirectlyToFile ? FileAccess.ReadWrite : FileAccess.Read,
+                            FileShare.Read,
+                            callFlushFileBuffers: false);
+                    }
+
+                    this.dataFileHandle.Seek(0, SeekOrigin.End);
+                    return;
+                }
+                catch (IOException e)
+                {
+                    lastException = e;
+                }
+                catch (UnauthorizedAccessException e)
+                {
+                    lastException = e;
+                }
+
+                if (retryUntilSuccess)
+                {
+                    if (this.Tracer != null && attempts % IoFailureLoggingThreshold == 0)
+                    {
+                        EventMetadata metadata = CreateEventMetadata(lastException);
+                        metadata.Add("attempts", attempts);
+                        this.Tracer.RelatedWarning(metadata, nameof(this.OpenOrCreateDataFile) + ": Failed to open data file stream ... retrying");
+                    }
+
+                    ++attempts;
+                    Thread.Sleep(IoFailureRetryDelayMS);
+                }
+                else
+                {
+                    throw lastException;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Writes data as UTF8 to dataFileHandle. fileLock will be acquired.
+        /// </summary>
+        private void WriteToDisk(byte prefix, TEntry element)
+        {
+            if (!this.collectionAppendsDirectlyToFile)
+            {
+                throw new InvalidOperationException(nameof(this.WriteToDisk) + " requires that collectionAppendsDirectlyToFile be true");
+            }
+
+            lock (this.fileLock)
+            {
+                using (BinaryWriter w = new BinaryWriter(this.dataFileHandle, Encoding.UTF8, true))
+                {
+                    w.Write(prefix);
+                    this.serializeEntry(w, element);
+                }
+
+                this.dataFileHandle.Flush();
+            }
+        }
+
+        /// <summary>
+        /// Reads entries from dataFileHandle, removing any data after the last NewLine ("\r\n"). Requires fileLock.
+        /// </summary>
+        private void RemoveLastEntryIfInvalid()
```

Comment:
Why is this method empty?  The reason for this method was to keep the file from becoming corrupted by a power failure or a BSOD.  When a write was not flushed completely to disk, zeros would be at the end of the file which would cause issues loading the list.  To make this change resilient to BSODs then it would need to have an end marker for each entry so that this code find that last good entry and "fix" the file. 

Did you do any profiling to see where the time was spent when loading the placeholders?  Perhaps this method is the offender and it was searching for the last `\r\n` when it didn't need to?  I will try and find some time to profile and see if this is the case.

One other thought is if we should use a SQLite database for this instead of a custom binary format.  We are using one for the BlobSizes and have discussed moving other "databases" to use it.  Using it of course would be a question of performance and making sure transactions get committed and the database doesn't become corrupt during power loss or BSOD.

Reply:
Yep, you're right, for power failure (and other critical failure) scenarios, we need to implement this method. 

I didn't do any proper profiling (couldn't find a dotTrace equivalent for Mac), but this method, implemented properly, should read just the last few bytes (EOL marker) from the end of file, so should be fast. Working with strings is not.

Yes, we could standardize on SQLite, that would be a more maintainable format. I didn't see, that it's used elsewhere and didn't want to introduce new dependencies. 

How are you dealing with a potential corruption of the SQLite, that's being used for BlobSizes?

Reply:
@wilbaker Did the work around the BlobSizes which can be found [here](https://github.com/Microsoft/VFSForGit/blob/master/GVFS/GVFS.Virtualization/BlobSize/BlobSizes.cs).  The main difference with BlobSizes is that if the write doesn't make it to the database it will download the information again and try to write again.  The placeholders database should be more fault tolerant although we are losing a placeholder right now with the current implementation in critical failure scenarios so we would be no worse off than today.  We haven't had any cases where the SQLite database itself becomes corrupt so it seems very resilient to failure.  It has [extensive testing](https://sqlite.org/testing.html) around these scenarios which is good.

Reply:
> The placeholders database should be more fault tolerant although we are losing a placeholder right now with the current implementation in critical failure scenarios so we would be no worse off than today.

What @kewillford mentions here is key.  

In the case of the BlobSizes database we don't *really* care if entries never make it to the database, as that database is just a local cache of file sizes.  If writes are not completely flushed we can always request the file sizes again (this is what allows us to make the performance optimization of writing sizes in batches in a background thread).  

If we decide to use SQLite for the placeholders we'll need to be more strict.  From the perspective of the thread that's adding a placeholder, the writes need to be synchronous and flushed to disk before the thread returns from its callback.

> we are losing a placeholder right now with the current implementation in critical failure scenarios so we would be no worse off than today.

Please also note that the issue @kewillford mentioned here can be fixed by writing to the database before writing the placeholder (#442 is open to fix this).  

We should make sure that whatever option we go with for improving perf is at least as reliable as what we have today and does not prevent us from fixing the power-loss scenario mentioned above.

Reply:
I've implemented RemoveLastEntryIfInvalid method for BinaryFileBasedCollection. Unfortunately, trying it out, whether SQLite would perform adequately, is a bit harder than, I originally thought and I don't have the time to do this now. 

Are you guys open to merging this now, to benefit from the performance gains and then perhaps standardise on SQLite later? If so, we need to discuss disk layout migrations.